File: flowersec-go/internal/protocolv4/cbor_text_reference_test.go
package protocolv4

// Test-only signed DNS/host/Origin syntax. Passing this reference does not
// authorize actual DNS, TLS, HTTP admission, origin headers or provider behavior.
import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"net/netip"
	"os"
	"regexp"
	"strconv"
	"strings"
	"testing"
)

type cborTextReference struct {
	*cborReference
	idna *idnaReference
}

func newCBORTextReference(t testing.TB) *cborTextReference {
	r := newCBORReference(t)
	return &cborTextReference{cborReference: r, idna: newIDNAReference(t, r)}
}

var numericHost = regexp.MustCompile(`^[0-9.]+$`)
var numericLastLabel = regexp.MustCompile(`^(?:[0-9]+|0x[0-9a-f]*)$`)
var originPattern = regexp.MustCompile(`^([a-z][a-z0-9+.-]*)://(\[[0-9a-f:]+\]|[^:/?#@\\\[\]]+)(?::([0-9]+))?$`)
var originPort = regexp.MustCompile(`^(?:0|[1-9][0-9]{0,4})$`)

func forbiddenHostRune(cp rune) bool {
	// The explicit host separator/whitespace set does not depend on Go's
	// current Unicode property tables. U+0085 remains an IDNA validity error.
	if cp >= 9 && cp <= 13 || cp >= 0x2000 && cp <= 0x200a {
		return true
	}
	switch cp {
	case ' ', '[', ']', '%', '\\', '/', '?', '#', '@', 0xa0, 0x1680, 0x2028, 0x2029, 0x202f, 0x205f, 0x3000, 0xfeff:
		return true
	}
	return false
}

func ipv6Hex(addr netip.Addr) string {
	data := addr.As16()
	words, text := [8]uint16{}, make([]string, 8)
	for i := range words {
		words[i] = binary.BigEndian.Uint16(data[i*2 : i*2+2])
		text[i] = strconv.FormatUint(uint64(words[i]), 16)
	}
	best, length := -1, 1
	for i := 0; i < len(words); {
		if words[i] != 0 {
			i++
			continue
		}
		end := i + 1
		for end < len(words) && words[end] == 0 {
			end++
		}
		if end-i > length {
			best, length = i, end-i
		}
		i = end
	}
	if best < 0 {
		return strings.Join(text, ":")
	}
	return strings.Join(text[:best], ":") + "::" + strings.Join(text[best+length:], ":")
}

func (r *cborTextReference) issuerHost(input string) (string, error) {
	if input == "" {
		return "", cborRefError("host_text")
	}
	if strings.ContainsFunc(input, forbiddenHostRune) {
		return "", cborRefError("host_syntax")
	}
	if strings.Contains(input, ":") {
		addr, err := netip.ParseAddr(input)
		if err != nil || !addr.Is6() {
			return "", cborRefError("host_ipv6")
		}
		// Preserve IPv4-mapped address family and all bits; never Unmap.
		return ipv6Hex(addr), nil
	}
	if numericHost.MatchString(input) {
		addr, err := netip.ParseAddr(input)
		if err != nil || !addr.Is4() {
			return "", cborRefError("host_ipv4")
		}
		return addr.String(), nil
	}
	dns, err := r.idna.issuerDNS(input)
	if err != nil {
		return "", err
	}
	if numericLastLabel.MatchString(dns[strings.LastIndexByte(dns, '.')+1:]) {
		return "", cborRefError("host_numeric_final_label")
	}
	return dns, nil
}

func (r *cborTextReference) wireHost(input string) error {
	if input == "" || strings.ContainsFunc(input, func(cp rune) bool { return cp >= 128 }) {
		return cborRefError("host_wire_ascii")
	}
	canonical, err := r.issuerHost(input)
	if err != nil {
		return err
	}
	if canonical != input {
		return cborRefError("host_noncanonical")
	}
	return nil
}

func (r *cborTextReference) originDefaultPort(scheme string) (uint64, error) {
	entries, err := r.fieldRegistry("origin_schemes")
	if err != nil {
		return 0, err
	}
	raw, ok := entries[scheme]
	if !ok {
		return 0, cborRefError("origin_scheme_unregistered")
	}
	var entry struct {
		Port *uint64 `json:"default_port"`
	}
	if json.Unmarshal(raw, &entry) != nil || entry.Port == nil {
		return 0, cborRefError("registry_unresolved")
	}
	return *entry.Port, nil
}

func (r *cborTextReference) wireOrigin(input string) error {
	if input == "" || strings.ContainsFunc(input, func(cp rune) bool { return cp < 0x21 || cp > 0x7e }) {
		return cborRefError("origin_ascii")
	}
	parts := originPattern.FindStringSubmatch(input)
	if parts == nil {
		return cborRefError("origin_syntax")
	}
	defaultPort, err := r.originDefaultPort(parts[1])
	if err != nil {
		return err
	}
	host := parts[2]
	if strings.HasPrefix(host, "[") {
		host = host[1 : len(host)-1]
		if !strings.Contains(host, ":") {
			return cborRefError("origin_syntax")
		}
	}
	if err := r.wireHost(host); err != nil {
		return err
	}
	if portText := parts[3]; portText != "" {
		port, err := strconv.ParseUint(portText, 10, 16)
		if err != nil || !originPort.MatchString(portText) {
			return cborRefError("origin_port")
		}
		if port == defaultPort {
			return cborRefError("origin_default_port")
		}
	}
	return nil
}

func (r *cborTextReference) textFormat(format, value string) error {
	switch format {
	case "host":
		return r.wireHost(value)
	case "origin":
		return r.wireOrigin(value)
	case "loopback_host":
		if err := r.wireHost(value); err != nil {
			return err
		}
		addr, err := netip.ParseAddr(value)
		if err != nil || !(value == "::1" || addr.Is4() && addr.As4()[0] == 127) {
			return cborRefError("host_loopback")
		}
		return nil
	}
	return cborRefError("text_format_unresolved")
}

func (r *cborTextReference) fieldFormats(field *cborRefField, value *cborRefValue, context cborShapeContext) error {
	field, err := r.resolveShapeField(field, context)
	if err != nil {
		return err
	}
	if field.TextFormat != "" {
		if value.major != 3 {
			return cborRefError("field_type")
		}
		return r.textFormat(field.TextFormat, string(value.data))
	}
	switch field.Type {
	case "array":
		for _, item := range value.items {
			if err := r.fieldFormats(field.Items, item, context); err != nil {
				return err
			}
		}
	case "text_map":
		for _, pair := range value.pairs {
			if err := r.fieldFormats(field.Keys, pair[0], context); err != nil {
				return err
			}
			child := field.Values
			if field.Entries != nil {
				child = field.Entries[string(pair[0].data)]
			}
			if err := r.fieldFormats(child, pair[1], context); err != nil {
				return err
			}
		}
	}
	// Map/embedded document children are visited by the common rule walker.
	return nil
}

func (r *cborTextReference) checkTextMap(name string, value *cborRefValue, context cborShapeContext) error {
	for _, pair := range value.pairs {
		field := r.registry.Maps[name].Fields[strconv.FormatUint(pair[0].n, 10)]
		if err := r.fieldFormats(field, pair[1], context); err != nil {
			return err
		}
	}
	if err := r.checkRelations(name, value, context); err != nil {
		return err
	}
	for _, raw := range r.registry.TextRules[name] {
		var rule struct {
			Op, Field, Format, Host, Port, Origin, Scheme string
			When                                          *cborRuleCondition
		}
		decoder := json.NewDecoder(bytes.NewReader(raw))
		decoder.DisallowUnknownFields()
		if decoder.Decode(&rule) != nil {
			return cborRefError("rule_unresolved")
		}
		applies, err := r.ruleApplies(name, value, rule.When, context)
		if err != nil {
			return err
		}
		if !applies {
			continue
		}
		switch rule.Op {
		case "text_format":
			v, err := r.variantPath(name, value, rule.Field, context)
			if err != nil {
				return err
			}
			if v == nil || v.major != 3 {
				return cborRefError("field_type")
			}
			if err := r.textFormat(rule.Format, string(v.data)); err != nil {
				return err
			}
		case "origin_endpoint":
			fields := make([]*cborRefValue, 0, 3)
			for _, field := range []string{rule.Host, rule.Port, rule.Origin} {
				v, err := r.variantPath(name, value, field, context)
				if err != nil || v == nil {
					return cborRefError("unknown_rule_field")
				}
				fields = append(fields, v)
			}
			if fields[0].major != 3 || fields[1].major != 0 || fields[2].major != 3 {
				return cborRefError("field_type")
			}
			host := string(fields[0].data)
			if strings.Contains(host, ":") {
				host = "[" + host + "]"
			}
			defaultPort, err := r.originDefaultPort(rule.Scheme)
			if err != nil {
				return err
			}
			expected := rule.Scheme + "://" + host
			if fields[1].n != defaultPort {
				expected += ":" + strconv.FormatUint(fields[1].n, 10)
			}
			if expected != string(fields[2].data) {
				return cborRefError("origin_endpoint")
			}
		default:
			return cborRefError("rule_unresolved")
		}
	}
	return nil
}

func (r *cborTextReference) wireMap(input []byte, name string, context cborShapeContext, cap uint64) (*cborRefValue, error) {
	value, err := r.shape(input, name, context, cap)
	if err != nil {
		return nil, err
	}
	if name != "" {
		if err := r.walkRuleMap(name, value, context, r.checkTextMap); err != nil {
			return nil, err
		}
	}
	return value, nil
}

// v4.cbor.text_corpus
func TestCBORTextReferenceCorpus(t *testing.T) {
	r := newCBORTextReference(t)
	raw, err := os.ReadFile("../../../testdata/transport_v4/text.json")
	if err != nil {
		t.Fatal(err)
	}
	var corpus struct {
		Vectors []struct {
			ID, Operation, Input, Output string
			ExpectedError                string `json:"expected_error"`
			Hex                          string `json:"utf8_hex"`
		}
	}
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatal(err)
	}
	for _, vector := range corpus.Vectors {
		t.Run(vector.ID, func(t *testing.T) {
			output, err := vector.Input, error(nil)
			switch vector.Operation {
			case "issuer_dns":
				output, err = r.idna.issuerDNS(vector.Input)
			case "wire_dns":
				err = r.idna.wireDNS(vector.Input)
			case "issuer_host":
				output, err = r.issuerHost(vector.Input)
			case "wire_host":
				err = r.wireHost(vector.Input)
			case "wire_origin":
				err = r.wireOrigin(vector.Input)
			default:
				t.Fatal("unknown text operation")
			}
			if vector.ExpectedError != "" {
				if err == nil {
					t.Fatal("accepted invalid text")
				}
				return
			}
			if err != nil || output != vector.Output || hex.EncodeToString([]byte(output)) != vector.Hex {
				t.Fatalf("text result %q/%v, want %q", output, err, vector.Output)
			}
		})
	}
	if len(corpus.Vectors) == 0 {
		t.Fatal("empty text coverage")
	}
	t.Logf("all %d shared issuer/wire text cases passed", len(corpus.Vectors))
}

// v4.cbor.wire_maps
func TestCBORWireMapReferenceCorpus(t *testing.T) {
	r := newCBORTextReference(t)
	positive, negative := 0, 0
	for _, vector := range cborRefVectors(t) {
		if vector.ExpectedError == "pool_set_membership" || vector.ExpectedError == "open_digest_mismatch" {
			continue
		}
		t.Run(vector.ID, func(t *testing.T) {
			input, err := hex.DecodeString(vector.Hex)
			if err != nil {
				t.Fatal(err)
			}
			original := bytes.Clone(input)
			value, err := r.wireMap(input, vector.Schema, shapeContext(vector.Limits), uint64(len(input))+1)
			if !bytes.Equal(input, original) {
				t.Fatal("wire validation changed original input")
			}
			if vector.ExpectedError != "" {
				negative++
				if err == nil || value != nil {
					t.Fatal("accepted invalid wire map")
				}
				return
			}
			positive++
			if err != nil || value == nil || !bytes.Equal(value.encode(nil), input) {
				t.Fatalf("wire map result %v", err)
			}
		})
	}
	if positive == 0 || negative == 0 {
		t.Fatal("empty wire-map coverage")
	}
	t.Logf("%d positives and %d negatives; external pool/open-digest oracles excluded", positive, negative)
}

// v4.cbor.text_fuzz
func FuzzCBORTextReference(f *testing.F) {
	r := newCBORTextReference(f)
	for _, seed := range []string{"example.com", "xn--bcher-kva.example", "\u4f8b\u5b50.\u6d4b\u8bd5", "\u0915\u094d\u200d\u0937.example", "a1.\u0645\u062b\u0627\u0644", "xn--zzzzzzzzzzzzzzzzzzzzzz", "EXAMPLE.COM", "127.0.0.1", "::ffff:192.0.2.1", "2001:db8::1"} {
		f.Add(seed)
	}
	f.Fuzz(func(t *testing.T, input string) {
		if len(input) > 1024 {
			return
		}
		if ascii, err := r.idna.issuerDNS(input); err == nil {
			if err := r.idna.wireDNS(ascii); err != nil {
				t.Fatalf("issuer DNS output rejected by wire rules: %q/%v", ascii, err)
			}
		}
		if canonical, err := r.issuerHost(input); err == nil {
			if err := r.wireHost(canonical); err != nil {
				t.Fatalf("issuer host output rejected by wire rules: %q/%v", canonical, err)
			}
			if err := r.wireOrigin("https://" + func() string {
				if strings.Contains(canonical, ":") {
					return "[" + canonical + "]"
				}
				return canonical
			}()); err != nil {
				t.Fatalf("canonical host failed Origin construction: %q/%v", canonical, err)
			}
		}
	})
}

File: flowersec-ts/src/v4/cborTextReference.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { transportV4Registry } from "../generated/transportV4Registry.js";
import { encode, utf8 } from "./testSupport/cbor.js";
import type { Value } from "./testSupport/cbor.js";
import { integer } from "./testSupport/shape.js";
import type { Context } from "./testSupport/shape.js";
import { hex } from "./testSupport/rules.js";
import { scalar, scalarText } from "./testSupport/idna.js";
import { TextReference, ipv6, ipv6Hex } from "./testSupport/text.js";
import { V4Failure, root } from "./testSupport/unicode.js";

type Vector = { id: string; hex: string; schema?: string; expected_error?: string; limits?: Record<string, number | string> };
const corpus = JSON.parse(readFileSync(new URL("testdata/transport_v4/corpus.json", root), "utf8")) as { schema_sha256: string; vectors: Vector[] };
const reference = new TextReference();
const context = (vector: Vector): Context => ({
  limits: Object.fromEntries(Object.entries(vector.limits ?? {}).filter(([, value]) => typeof value === "number").map(([key, value]) => [key, integer(value)])),
  selectors: Object.fromEntries(Object.entries(vector.limits ?? {}).filter((entry): entry is [string, string] => typeof entry[1] === "string")),
});

describe("v4.ts_text", () => {
  // v4.ts_text.corpus
  it("consumes every shared issuer and wire text vector", () => {
    type TextVector = { id: string; operation: string; input: string; output?: string; utf8_hex?: string; expected_error?: string };
    const corpus = JSON.parse(readFileSync(new URL("testdata/transport_v4/text.json", root), "utf8")) as { schema_sha256: string; vectors: TextVector[] };
    expect(corpus.schema_sha256).toBe(transportV4Registry.schemaSHA256);
    for (const vector of corpus.vectors) {
      const operation = (): string => {
        switch (vector.operation) {
          case "issuer_dns": return reference.idna.issuerDNS(vector.input);
          case "wire_dns": reference.idna.wireDNS(vector.input); break;
          case "issuer_host": return reference.issuerHost(vector.input);
          case "wire_host": reference.wireHost(vector.input); break;
          case "wire_origin": reference.wireOrigin(vector.input); break;
          default: throw new Error("unknown text operation");
        }
        return vector.input;
      };
      if (vector.expected_error) expect(operation, vector.id).toThrow(V4Failure);
      else {
        const output = operation(); expect(output, vector.id).toBe(vector.output); expect(utf8.encode(output), vector.id).toEqual(hex(vector.utf8_hex));
      }
    }
    expect(corpus.vectors.length).toBe(100);
  });

  // v4.ts_text.wire_maps
  it("validates the complete map corpus except external composition", () => {
    expect(corpus.schema_sha256).toBe(transportV4Registry.schemaSHA256);
    let positive = 0, negative = 0, separate = 0;
    for (const vector of corpus.vectors) {
      if (["pool_set_membership", "open_digest_mismatch"].includes(vector.expected_error ?? "")) { separate += 1; continue; }
      const raw = hex(vector.hex), ctx = context(vector), before = structuredClone(ctx);
      const decode = (): Value => reference.wireMap(raw, vector.schema, ctx, BigInt(raw.length) + 1n);
      if (vector.expected_error) { expect(decode, vector.id).toThrow(V4Failure); negative += 1; }
      else { expect(encode(decode()), vector.id).toEqual(raw); positive += 1; }
      expect(ctx).toEqual(before);
    }
    expect(positive).toBe(337); expect(negative).toBe(826); expect(separate).toBe(2);
  });

  // v4.ts_text.address_boundaries
  it("preserves address family/bits and requires canonical registered Origins", () => {
    for (const [input, canonical] of [["2001:0DB8:0:0:1:0:0:1", "2001:db8::1:0:0:1"], ["0:0:0:0:0:0:0:0", "::"], ["::FFFF:192.0.2.1", "::ffff:c000:201"], ["0:0:0:0:0:0:0:1", "::1"]]) {
      expect(reference.issuerHost(input!)).toBe(canonical); expect(ipv6(input!)).toEqual(ipv6(canonical!));
      reference.wireHost(canonical!); expect(() => reference.wireHost(input!)).toThrow(V4Failure);
    }
    for (const host of ["127.1", "127.00.0.1", "2130706433", "0x7f000001", "example.0x", "example.123", "::ffff:127.0.0.1%lo", "example.com.", "example.com/path", "example.com\u00a0", "1::2::3", ":::1", "1:::2", "1:2:3:4:5:6:7", "1:2:3:4:5:6:7:8:9", "192.0.2.1::", "+1::", "::ffff:192.00.2.1"]) expect(() => reference.issuerHost(host), host).toThrow(V4Failure);
    reference.wireHost("example.0xg");
    for (const origin of ["https://example.com", "https://example.com:0", "https://[::ffff:c000:201]", "http://127.0.0.1:3000"]) reference.wireOrigin(origin);
    for (const origin of ["https://example.com:443", "http://127.0.0.1:80", "https://example.com:0443", "https://example.com:65536", "https://user@example.com", "https://example.com/", "https://example.com?x", "https://example.com#x", "https://EXAMPLE.COM", "https://[::ffff:192.0.2.1]", "https://example.com\n", "unregistered://example.com", "https://example.com:", "https://[example.com]", "https://[::1]]", "https://[::1]:+1"]) expect(() => reference.wireOrigin(origin), origin).toThrow(V4Failure);
    expect(ipv6Hex([1, 0, 0, 2, 0, 0, 3, 4])).toBe("1::2:0:0:3:4");
    expect(ipv6Hex([1, 0, 2, 3, 4, 5, 6, 7])).toBe("1:0:2:3:4:5:6:7");
    expect(ipv6("::ffff:c000:201")).toEqual([0, 0, 0, 0, 0, 65535, 49152, 513]);
  });

  // v4.ts_text.properties
  it("preserves address/issuer identity and mutated maps over three 4096-case paths", () => {
    let state = 0x41e7141e;
    function next(): number { state ^= state << 13; state ^= state >>> 17; state ^= state << 5; return state >>> 0; }
    for (let iteration = 0; iteration < 4096; iteration++) {
      const words = Array.from({ length: 8 }, () => { const n = next(); return n % 3 === 0 ? 0 : n & 0xffff; });
      const full = words.map(word => word.toString(16).toUpperCase().padStart(4, "0")).join(":"), canonical = reference.issuerHost(full);
      expect(ipv6(canonical)).toEqual(words); reference.wireHost(canonical); reference.wireOrigin(`https://[${canonical}]`);
      const input = scalarText(Array.from({ length: next() % 128 }, () => next() % 0x110000).filter(scalar));
      let host: string | undefined;
      try { host = reference.issuerHost(input); } catch (error) { expect(error).toBeInstanceOf(V4Failure); }
      if (host !== undefined) { reference.wireHost(host); reference.wireOrigin("https://" + (host.includes(":") ? `[${host}]` : host)); }
      const vector = corpus.vectors[next() % corpus.vectors.length]!, raw = hex(vector.hex), ctx = context(vector), before = structuredClone(ctx);
      if (raw.length) { const at = next() % raw.length; raw[at] = raw[at]! ^ (next() & 255); }
      let value: Value | undefined;
      try { value = reference.wireMap(raw, vector.schema, ctx, BigInt(raw.length) + 1n); } catch (error) { expect(error).toBeInstanceOf(V4Failure); }
      if (value) expect(encode(value)).toEqual(raw); expect(ctx).toEqual(before);
    }
  }, 30000);
});
